File: +postprocess/coherent_compounding.m
classdef coherent_compounding < postprocess
%COHERENT_COMPOUNDING   Matlab implementation of Coherent compounding
% If the window_size property is specified, a sliding window compounding approach with window size
% window_size is performed. Sliding window compounding is only performed along the transmit
% dimension. The receive dimension is compounded as usual
%
%   authors: Alfonso Rodriguez-Molares (alfonso.r.molares@ntnu.no)
%            Ole Marius Hoel Rindal <olemarius@olemarius.net>
%            Fabrice Prieur
%            Stefano Fiorentini <stefano.fiorentini@ntnu.no>
%
%   $Last updated: 2023/09/01$

    %% constructor
    methods (Access = public)
        function h=coherent_compounding()
            h.name='Coherent compounding MATLAB';   
            h.reference='www.ntnu.no';                
            h.implemented_by={'Alfonso Rodriguez-Molares <alfonso.r.molares@ntnu.no>', ...
                'Ole Marius Hoel Rindal <olemarius@olemarius.net>', ...
                'Fabrice Prieur', ...
                'Stefano Fiorentini <stefano.fiorentini@ntnu.no>'};   
            h.version='v1.1.0';
        end
    end
    
    properties (Access = public)
       dimension = dimension.both;          % Which "dimension" to sum over
       window_size = []                     % If specified, do moving sum along transmit dimension
    end
    
    methods
        function output=go(h)
            % check if we can skip calculation
            if h.check_hash()
                output= h.output; 
                return;
            end
            
                        
            % Define output object
            h.output=uff.beamformed_data(h.input);
            
            if isempty(h.window_size)
                switch h.dimension
                    case dimension.both %#ok<*PROP>
                        h.output.data=sum(h.input.data,[2,3]);
                    case dimension.transmit
                        h.output.data=sum(h.input.data,3);
                    case dimension.receive
                        h.output.data=sum(h.input.data,2);
                    otherwise
                        error('Unknown dimension mode; check HELP dimension');
                end
            else
                assert(h.window_size <= h.input.N_waves, ...
                'Cannot have a window size greater that the available number of transmit events.')
            
                data = reshape(h.input.data, [h.input.N_pixels, h.input.N_channels, 1, h.input.N_waves*h.input.N_frames]);
                
                switch h.dimension
                    case dimension.both
                        h.output.data = movsum(sum(data, 2), h.window_size, 4, 'Endpoints', 'discard');
                    case dimension.transmit
                        h.output.data=movsum(data, h.window_size, 4, 'Endpoints', 'discard');
                    case dimension.receive
                        h.output.data=sum(h.input.data,2);
                    otherwise
                        error('Unknown dimension mode; check HELP dimension');
                end
            end            
            
            % pass reference
            output = h.output;
            
            % update hash
            h.save_hash();

        end
    end
    
    methods
        function set.dimension(h, val)
            validateattributes(val, {'dimension'}, {'scalar'})
            h.dimension = val;
        end
        function set.window_size(h, val)
            if ~isempty(val)
                validateattributes(val, {'single', 'double'}, {'integer', 'scalar'})
            end
            h.window_size = val;
        end
    end
end

File: python/src/ustb/postprocess/coherent_compounding.py
"""Coherent compounding postprocess matching MATLAB postprocess.coherent_compounding.

Sums complex beamformed data across the transmit/receive dimension(s),
preserving phase (as opposed to incoherent compounding, which sums
magnitudes).
"""

import numpy as np
from numpy.lib.stride_tricks import sliding_window_view

from ustb.enums import Dimension
from ustb.beamformed_data import BeamformedData


def _moving_sum(x, window_size, axis):
    """Sliding-window sum along ``axis``, discarding incomplete windows.

    Matches MATLAB movsum(x, window_size, axis, 'Endpoints', 'discard'):
    output length along axis is N - window_size + 1.
    """
    x = np.moveaxis(x, axis, -1)
    windows = sliding_window_view(x, window_shape=window_size, axis=-1)
    summed = windows.sum(axis=-1)
    return np.moveaxis(summed, -1, axis)


class CoherentCompounding:
    """Coherent compounding postprocess matching MATLAB postprocess.coherent_compounding."""

    def __init__(self):
        self.input = None
        self.dimension = Dimension.both
        self.window_size = None

    def go(self):
        data = np.asarray(self.input.data)
        while data.ndim < 4:
            data = data[..., np.newaxis]

        if self.window_size is None:
            if self.dimension == Dimension.both:
                out = np.sum(data, axis=(1, 2), keepdims=True)
            elif self.dimension == Dimension.transmit:
                out = np.sum(data, axis=2, keepdims=True)
            elif self.dimension == Dimension.receive:
                out = np.sum(data, axis=1, keepdims=True)
            else:
                raise ValueError(f"Unsupported dimension: {self.dimension}")
        else:
            N_pixels, N_channels, N_waves, N_frames = data.shape
            if self.window_size > N_waves:
                raise ValueError(
                    "Cannot have a window size greater than the available "
                    "number of transmit events."
                )
            reshaped = data.reshape(N_pixels, N_channels, 1, N_waves * N_frames)

            if self.dimension == Dimension.both:
                summed = np.sum(reshaped, axis=1, keepdims=True)
                out = _moving_sum(summed, self.window_size, axis=3)
            elif self.dimension == Dimension.transmit:
                out = _moving_sum(reshaped, self.window_size, axis=3)
            elif self.dimension == Dimension.receive:
                # Matches MATLAB: window_size is ignored for dimension.receive.
                out = np.sum(data, axis=1, keepdims=True)
            else:
                raise ValueError(f"Unsupported dimension: {self.dimension}")

        return BeamformedData(scan=self.input.scan, data=out.astype(data.dtype))
